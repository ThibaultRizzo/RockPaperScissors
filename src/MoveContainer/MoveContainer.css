.MoveContainer {
  display: flex;
  flex-flow: row nowrap;
  justify-content: space-around;
  align-items: center;
}

.Move {
  display: flex;
  justify-content: center;
  align-items: center;
  width: 150px;
  height: 150px;
  border-radius: 50%;
  background: rgb(154, 101, 23);
  background: radial-gradient(
    circle,
    rgba(255, 182, 74, 1) 41%,
    rgba(255, 226, 175, 1) 100%
  );
  box-shadow: rgba(255, 226, 175, 1) 0px 0px 1px 2px;
  transition: all 0.5s;
}

.Move:hover {
  box-shadow: rgba(255, 226, 175, 1) 0px 0px 1px 20px;
}

.Move > input {
  border: none;
  width: 60%;
  height: 60%;
  background-size: 100%;
  background-clip: content-box;
  background-origin: content-box;
  transition: all 0.5s;
}

.Move > input:focus {
  outline: none;
}

.Move > input:hover {
  width: 70%;
  height: 70%;
}

.game-type {
  flex-basis: 100%;
  text-align: center;
  color: #8c7272;
  font-style: italic;
  font-size: 50px;
  margin: 0;
}

.number-turn {
  flex-basis: 100%;
  font-size: 30px;
  color: #544343;
  text-align: center;
}

input[name='SCISSORS'] {
  background: url('./../resources/scissors.svg') no-repeat;
}

input[name='PAPER'] {
  background: url('./../resources/paper.svg') no-repeat;
}

input[name='ROCK'] {
  background: url('./../resources/rock.svg') no-repeat;
}
