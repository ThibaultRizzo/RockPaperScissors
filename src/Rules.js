/**
 * Enum object of all possible moves possible.
 */
const GameMoves = {
  ROCK: 'ROCK',
  PAPER: 'PAPER',
  SCISSORS: 'SCISSORS'
};

/**
 * Enum object of all game types
 */
const GameType = {
  PlayerVsAI: 'Player vs AI',
  AIVsAI: 'AI vs AI'
};

/**
 * Enum object of possible outcomes in a game.
 */
const Result = {
  DRAW: 'DRAW',
  WIN: 'WIN',
  LOSE: 'LOSE'
};

/**
 * Enum object encapsulating all game rules
 * There are redundant information displayed for clarity and enforcing coherence
 */
const GameRules = {
  ROCK: {
    winOver: [GameMoves.SCISSORS],
    loseOver: [GameMoves.PAPER]
  },
  PAPER: {
    winOver: [GameMoves.ROCK],
    loseOver: [GameMoves.SCISSORS]
  },
  SCISSORS: {
    winOver: [GameMoves.PAPER],
    loseOver: [GameMoves.ROCK]
  }
};

// Enforces enumerable principles by forbidding to modify the objet
Object.freeze(GameRules);
Object.freeze(GameMoves);
Object.freeze(GameType);
Object.freeze(Result);

export { GameRules, GameMoves, Result, GameType };

/**
 * Applies rules of game as specified in the GameRule object
 * Returns a result of type Result else null if one of the moves is not defined
 * @param {*} p1: a value from GameMoves
 * @param {*} p2: a value from GameMoves
 */
export function winner(p1, p2) {
  let result;
  // Check that both objects are of GameMoves type
  if (
    !Object.values(GameMoves).includes(p1) &&
    !Object.values(GameMoves).includes(p2)
  ) {
    return null;
  }
  if (p1 === p2) {
    result = Result.DRAW;
  } else if (GameRules[p1].winOver.includes(p2)) {
    result = Result.WIN;
  } else {
    result = Result.LOSE;
  }
  return result;
}

export function generateRandomMove() {
  const randomPosition = Math.floor(
    Math.random() * Object.values(GameMoves).length
  );
  const result = Object.values(GameMoves)[randomPosition];
  return result;
}

/**
 * Resolve the turn given player move
 * Returns moves of both players and result if move of P1 is defined
 * Else returns nulled properties
 * @param {*} moveP1
 */
export function resolveTurn(moveP1) {
  let turn = {
    moveP1: moveP1,
    moveP2: null,
    result: null
  };
  if (turn.moveP1) {
    turn.moveP2 = generateRandomMove();
    turn.result = winner(turn.moveP1, turn.moveP2);
  }
  return turn;
}

/**
 * Returns game object with a details properties summarizing the game.
 * ex:
 *      details: {
 *          result: -> Result enum object
 *          scoreP1: -> Number of turn won by P1
 *          scoreP2: -> Number of turn won by P2
 *      }
 * @param {*} game
 */
export function summarizeGame(game, gameType) {
  let gameResult = {
    result: null,
    scoreP1: 0,
    scoreP2: 0,
    gameType: gameType,
  };
  game.forEach(turn => {
    if (Result.WIN === turn.result) {
      gameResult.scoreP1++;
    } else if (Result.LOSE === turn.result) {
      gameResult.scoreP2++;
    }
  });
  if (gameResult.scoreP1 > gameResult.scoreP2) {
    gameResult.result = Result.WIN;
  } else if (gameResult.scoreP1 < gameResult.scoreP2) {
    gameResult.result = Result.DRAW;
  } else {
    gameResult.result = Result.LOSE;
  }
  return gameResult;
}

/**
 * Calculates win rate of player given all games recorded, else return O
 * @param {*} games
 */
export function workWinRate(games) {
  let wins = 0;

  games.forEach(game => {
    let details = game[game.length - 1];
    if (details.result === Result.WIN) {
      wins++;
    }
  });

  let winRate =
    Object.keys(games).length !== 0 ? wins / Object.keys(games).length : 0;
  return Math.floor(winRate * 100) + '%';
}
