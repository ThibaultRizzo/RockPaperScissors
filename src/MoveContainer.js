import React, { Component } from 'react';
import Move from './Move.js';
import './MoveContainer.css';
import { GameMoves } from './Rules.js';

class MoveContainer extends Component {
  render() {
    const movesList = GameMoves.map(move => {
      return (
        <Move key={move} type={move} onMoveClicked={this.props.onMoveClicked} />
      );
    });
    const gameType = this.props.gameType ? this.props.gameType : false;
    return (
    <div id="MoveContainer">
      <h1 className="game-type">{gameType}</h1>
      {movesList}
    </div>
    );
  }
}

export default MoveContainer;
