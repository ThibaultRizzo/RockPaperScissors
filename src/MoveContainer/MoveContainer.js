import React, { Component } from 'react';
import Move from './Move.js';
import './MoveContainer.css';
import { GameMoves, generateRandomMove } from './../Rules.js';

class MoveContainer extends Component {
  render() {
    const movesList = [];
    for (let move in GameMoves) {
      movesList.push(
        <Move key={move} type={move} onMoveClicked={this.props.onMoveClicked} />
      );
    }

    const gameType = this.props.gameType ? this.props.gameType : false;
    if (gameType === 'AI vs AI' && !this.props.isGameFinished) {
      setTimeout(this.props.onMoveClicked(generateRandomMove()), 1000);
    }
    return (
      <div id="MoveContainer">
        <h1 className="game-type">{gameType}</h1>
        <h2 className="number-turn">
          Number of turns: {this.props.maxturnCount}
        </h2>
        {movesList}
      </div>
    );
  }
}

export default MoveContainer;
