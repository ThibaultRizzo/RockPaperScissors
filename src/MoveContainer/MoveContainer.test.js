import React from 'react';
import ReactDOM from 'react-dom';
import MoveContainer from './MoveContainer';
import renderer from 'react-test-renderer';
import { GameType } from '../Rules';

it('renders correctly', () => {
  const clicked = jest.fn();
  const startState = {
    isGameFinished: false,
    maxturnCount: 4,
    gameType: GameType.AIVsAI
  };
  const rendered = renderer.create(
    <MoveContainer
      isGameFinished={startState.isGameFinished}
      maxturnCount={startState.maxturnCount}
      gameType={startState.gameType}
      onMoveClicked={clicked}
    />
  );
  expect(rendered.toJSON()).toMatchSnapshot();
});
