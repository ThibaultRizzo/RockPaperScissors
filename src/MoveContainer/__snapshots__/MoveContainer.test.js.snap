// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  id="MoveContainer"
>
  <h1
    className="game-type"
  >
    AI vs AI
  </h1>
  <h2
    className="number-turn"
  >
    Number of turns: 
    4
  </h2>
  <div
    className="Move  "
  >
    <input
      name="ROCK"
      onClick={[Function]}
      type="button"
    />
  </div>
  <div
    className="Move  "
  >
    <input
      name="PAPER"
      onClick={[Function]}
      type="button"
    />
  </div>
  <div
    className="Move  "
  >
    <input
      name="SCISSORS"
      onClick={[Function]}
      type="button"
    />
  </div>
</div>
`;
